### PR TITLE
Rawtools: new Galaxy tool

### PR DESCRIPTION
**Pull request by CarlosHorro**
RawTools is an open-source and freely available package designed to perform scan data parsing and quantification, and quality control analysis of Thermo Orbitrap raw mass spectrometer files.

RawTools galaxy tool is on working condition, postponing the implementation of Quality Control analysis.

There are currently 2 main issues about which it would be great to have galaxy-p team feedback:

- Maximum size of a Galaxy tool. In order to develop tests, we need to include into the tool, at least, one .raw file and .mgf file; first ones are specially big (I have been reported that the minimum raw file available in my lab is about 66mb), and we do not know which could be an acceptable maximum size for a galaxy tool.
Current consequences of this doubt is tests have not been implemented yet (next issue implies another reason for this).

- We are having similar problems with raw files than the ones I had using moFF. When uploading our .raw files to Galaxy (my Galaxy version is 17_09, but we cannot update it for some time yet) , it shows an error with this message:

`WARNING:galaxy.model:Datatype class not found for extension 'raw'
`

I think that, in theory, thermo.raw file type supports .raw extension, but there is some problem with this extension that does not properly support our Thermo raw files with .raw extension.
I tried to upload the same file to usegalaxy.org and it also showed the same error, but the history item finished green instead of red... it would be great to have the same behavior (or no error at all, of course) here...
In order to bypass this problem on our server I had to install "proteomics_datatypes" tool, which adds "raw" extension support and it allows me to upload my .raw files successfully. 
But, this tool is deprecated and it seems that it is not currently the correct approach to solve this.
By the way, even with this tool installed and a "tool_dependencies.xml" file with proteomics_datatypes repo, I get an error when executing the first test I am trying to develop on my local (the one commented on the uploaded xml):

```
ERROR

======================================================================
ERROR: Raw Tools ( rawtools ) > Test-1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/parallels/GitHub/galaxy/test/functional/test_toolbox.py", line 317, in test_tool
    self.do_it(td)
  File "/home/parallels/GitHub/galaxy/test/functional/test_toolbox.py", line 48, in do_it
    stage_data_in_history(galaxy_interactor, testdef.test_data(), test_history, shed_tool_id)
  File "/home/parallels/GitHub/galaxy/test/base/interactor.py", line 44, in stage_data_in_history
    upload_wait()
  File "/home/parallels/GitHub/galaxy/test/base/interactor.py", line 314, in wait
    while not self.__history_ready(history_id):
  File "/home/parallels/GitHub/galaxy/test/base/interactor.py", line 336, in __history_ready
    return self._state_ready(state, error_msg="History in error state.")
  File "/home/parallels/GitHub/galaxy/test/base/interactor.py", line 406, in _state_ready
    raise Exception(error_msg)
Exception: History in error state.
-------------------- >> begin captured stdout << ---------------------
Problem in history with id 2891970512fa2d5a - summary of datasets below.
--------------------------------------
| 1 - qExactive01819.thermo.raw (HID - NAME) 
| Dataset State:
|  error
| Dataset Blurb:
|  error
| Dataset Info:
|  uploaded thermo.raw file
|  WARNING:galaxy.model:Datatype class not found for extension 'raw'
|  WARNING:galaxy.model:Datatype class not found for extension 'raw'
|  WARNING:galaxy.model:Datatype class not found for extension 'raw'
|  WARNING:galaxy.model:Datatype c
| Peek:
|  None
| Dataset Job Standard Output:
|  *Standard output was empty.*
| Dataset Job Standard Error:
|  WARNING:galaxy.model:Datatype class not found for extension 'raw'
|  WARNING:galaxy.model:Datatype class not found for extension 'raw'
|  WARNING:galaxy.model:Datatype class not found for extension 'raw'
|  WARNING:galaxy.model:Datatype class not found for extension 'raw'
|  WARNING:galaxy.model:Datatype class not found for extension 'raw'
|
--------------------------------------

--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
XML: /tmp/tmpo3rKaI/xunit.xml
----------------------------------------------------------------------
Ran 1 test in 27.905s

FAILED (errors=1)
2019-03-05 12:10:57,092 INFO  [test_driver] Shutting down
2019-03-05 12:10:57,093 INFO  [test_driver] Shutting down embedded galaxy web server
2019-03-05 12:10:57,097 INFO  [test_driver] Embedded web server galaxy stopped
2019-03-05 12:10:57,097 INFO  [test_driver] Stopping application galaxy
2019-03-05 12:10:58,594 INFO  [galaxy.jobs.handler] sending stop signal to worker thread
2019-03-05 12:10:58,594 INFO  [galaxy.jobs.handler] job handler queue stopped
2019-03-05 12:10:58,594 INFO  [galaxy.jobs.runners] TaskRunner: Sending stop signal to 2 worker threads
2019-03-05 12:10:58,595 INFO  [galaxy.jobs.runners] LocalRunner: Sending stop signal to 4 worker threads
2019-03-05 12:10:58,595 INFO  [galaxy.jobs.handler] sending stop signal to worker thread
2019-03-05 12:10:58,595 INFO  [galaxy.jobs.handler] job handler stop queue stopped
2019-03-05 12:10:58,596 INFO  [test_driver] Application galaxy stopped.
2019-03-05 12:10:58,597 INFO  [galaxy.jobs.handler] sending stop signal to worker thread
2019-03-05 12:10:58,598 INFO  [galaxy.jobs.handler] job handler queue stopped
2019-03-05 12:10:58,598 INFO  [galaxy.jobs.runners] TaskRunner: Sending stop signal to 2 worker threads
2019-03-05 12:10:58,598 INFO  [galaxy.jobs.runners] LocalRunner: Sending stop signal to 4 worker threads
2019-03-05 12:10:58,598 INFO  [galaxy.jobs.handler] sending stop signal to worker thread
2019-03-05 12:10:58,598 INFO  [galaxy.jobs.handler] job handler stop queue stopped
Testing complete. HTML report is in "/media/psf/shared/rawtools_for_tests/tool_test_output.html".
There were problems with 1 test(s) - out of 1 test(s) executed. See /media/psf/shared/rawtools_for_tests/tool_test_output.html for detailed breakdown.
Traceback (most recent call last):
  File "/home/parallels/Programs/Bio/planemo/.venv/bin/planemo", line 11, in <module>
    sys.exit(planemo())
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/planemo/cli.py", line 178, in handle_blended_options
    return f(*args, **kwds)
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/planemo/commands/cmd_test.py", line 77, in cli
    return_value = run_in_config(ctx, config, **kwds)
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/planemo/galaxy/test/actions.py", line 115, in run_in_config
    kwds=kwds
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/planemo/galaxy/test/actions.py", line 129, in handle_reports_and_summary
    **kwds
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/planemo/galaxy/test/actions.py", line 220, in _handle_summary
    **kwds
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/planemo/galaxy/test/actions.py", line 232, in _summarize_tests_full
    _summarize_test_case(test_case_data, **kwds)
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/planemo/galaxy/test/actions.py", line 250, in _summarize_test_case
    get_dict_value("data", structured_data)
  File "/home/parallels/Programs/Bio/planemo/.venv/local/lib/python2.7/site-packages/planemo/test/results.py", line 113, in get_dict_value
    raise KeyError("No key [%s] in [%s]" % (key, data))
KeyError: 'No key [status] in [None]'


```

So, at the end, this problem also prevents me to create proper tests...

I would be very thankful if someone could orientate us about the best way to focus these issues, please.

Greetings,
Carlos

### CODE REVIEW

**Comment by bgruening** [comment]
can you try thermo.raw see here: https://github.com/galaxyproject/galaxy/blob/dev/config/datatypes_conf.xml.sample#L234

**Comment by CarlosHorro** [comment]
> can you try thermo.raw see here: https://github.com/galaxyproject/galaxy/blob/dev/config/datatypes_conf.xml.sample#L234

Um... when changing the format to "thermo.raw" in the input field this problem still occurs...

```
<param format="thermo.raw" name="input_raw" type="data" label="Raw file" multiple="true"
            help="Select the raw file from history"/>
```

**Comment by eslerm** [comment]
@CarlosHorro, my lab shared a few very small Thermo files for this sort of thing: https://github.com/HegemanLab/small-thermo-raw-file-examples

We could add MSMS files if needed too.

**Comment by kevinkovalchik** [comment]
@eslerm, thanks for sharing the files. RawTools is specifically intended for data-dependent acquisition (DDA) .raw files from Orbitrap intruments (both MS2 and SPS-MS3 DDA is fine). Does your lab have any small files of that variety?

Kevin

File: tools/rawtools/rawtools.xml [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+       <xml name="input_fasta">
+         <param format="fasta" name="input_database" type="data" label="Protein Database"
+             help="Select FASTA database from history"/>
+       </xml>
+    </macros>
+    <requirements>
+        <requirement type="package" version="@RAWTOOLS_VERSION@">rawtools</requirement>
+      <!--  <repository changeset_revision="349af9298ff1" name="proteomics_datatypes" owner="iracooke" prior_installation_required="True" toolshed="https://testtoolshed.g2.bx.psu.edu" /> -->
+    </requirements>
+    <stdio>
```
Comment (bgruening) [inline tools/rawtools/rawtools.xml]: please replace this stdio with `detect_errors` on the command-element.

File: tools/rawtools/rawtools.xml [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+<![CDATA[
+        #from datetime import datetime
+        #import json
+        #import os
+        #set $exp_str = "Galaxy_Experiment_%s" % datetime.now().strftime("%Y%m%d%H%M%s")
+        #set $samp_str = "Sample_%s" % datetime.now().strftime("%Y%m%d%H%M%s")
+        #set $temp_stderr = "rawtools_stderr"
+        #set $bin_dir = "bin"
+
+        mkdir output;
```
Comment (bgruening) [inline tools/rawtools/rawtools.xml]: please use && instead of ;

File: tools/rawtools/rawtools.xml (hunk 1 above)
Comment (bgruening) [reply tools/rawtools/rawtools.xml]: https://docs.galaxyproject.org/en/latest/dev/schema.html#error-detection

File: tools/rawtools/rawtools.xml [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+        cwd=`pwd`;
+        export HOME=\$cwd;
```
Comment (bgruening) [inline tools/rawtools/rawtools.xml]: If you add profile="18.01" this should work without this hack.
https://github.com/galaxyproject/galaxy/pull/5193

File: tools/rawtools/rawtools.xml [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+            --out \$cwd/output
+
+        2>> $temp_stderr) &&
+
+        cat $temp_stderr 2>&1;
+]]>
+    </command>
+    <inputs>
+
+        <param format="raw" name="input_raw" type="data" label="Raw file" multiple="true"
```
Comment (bgruening) [inline tools/rawtools/rawtools.xml]: thermo.raw

File: tools/rawtools/rawtools.xml [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            help="Select the raw file from history"/>
+        <section name="parsing_quantification_options" expanded="true" title="Parsing and Quantification">
+            <conditional name="parsing_quantification_conditional">
+                <param name="matrix_boolean" type="boolean" truevalue="True" falsevalue="False" checked="true"
```
Comment (bgruening) [inline tools/rawtools/rawtools.xml]: please use a select here: https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html?highlight=boolean#booleans

File: tools/rawtools/rawtools.xml [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+                      help="Relative intensity cutoff for ions included in a MGF file. The value is relative to the highest intensity ion found in a given scan."/>
+                    -->
+                    <param name="mgf_refinement" label="Precursor mass and charge refinement" type="boolean" truevalue="True" falsevalue="False" checked="false"
+                      help="Refine precursor charge and monoisotopic mass assignments. Highly recommended if monoisotopic precursor selection was turned off in the instrument method (or peptide match on a QE instrument)."/>
+
+                </when>
+                <when value="False" />
+            </conditional>
+            <conditional name="chromatogram_conditional">
+                <param name="chromatogram_boolean" type="boolean" truevalue="True" falsevalue="False" checked="false"
```
Comment (bgruening) [inline tools/rawtools/rawtools.xml]: select

**Comment by eslerm** [comment]
@kevinkovalchik added https://github.com/HegemanLab/small-thermo-raw-file-examples/issues for this

I'll talk to the instrument operator.

**Comment by kevinkovalchik** [comment]
@eslerm Okay, thanks. I might be able to take care of it on my end by taking a slice out of a larger file. If I succeed would you like a copy to host with your other test files?

**Comment by kevinkovalchik** [comment]
@eslerm I was able to create a small file by slicing up a larger file.

**Comment by eslerm** [comment]
Nice @kevinkovalchik :)

How did you do it?

**Comment by kevinkovalchik** [comment]
There is a set of tools from Thermo called FT Programs, and in one of those (RecalOffline or something) you can save slices of raw files. It requires an installation of XCalibur to work.

File: tools/rawtools/rawtools.xml (hunk 3 above)
Comment (CarlosHorro) [reply tools/rawtools/rawtools.xml]: We are very interested on keeping support for Galaxy 17.09...

Comment (bgruening) [reply tools/rawtools/rawtools.xml]: Outch, this version does not even get security patches afaik. But ok :)

**Comment by bgruening** [comment]
Can you enable tests, I will have a look at this then.

**Comment by CarlosHorro** [comment]
> Can you enable tests, I will have a look at this then.

Yes, I'm working on them.

**Comment by CarlosHorro** [comment]
As RawTools can execute multiple files at the same time (and also return multiple files of the same type), I think I should use collections in all outputs.
I've been trying to start to do it with output_mgf, but I'm not able to do it. What is wrong if I do something like this? 

```
<collection name="output_mgf" type="list" label="${tool.name} - ${on_string}: MGF">
            <filter>pq_options['mgf_conditional']['mgf_select'] == "yes"</filter>
            <discover_datasets pattern="(\S+)\.mgf" ext="mgf" directory="output"/>
        </collection>
```

I think it should work, but the history item in Galaxy stays infinitely yellow waiting to find a dataset which fits to the discover_datasets params... 
MGF files generated finish with .mgf extension and the pattern is like the one specified there. This is the content of the "output" folder:

```
-rw-r--r-- 1 root root   2447597 Mar 25 12:22 ch_23Aug2018_HeLa_Std_1.raw_Matrix.txt
-rw-r--r-- 1 root root       737 Mar 25 12:22 ch_23Aug2018_HeLa_Std_1.raw_Metrics.txt
-rw-r--r-- 1 root root 137754912 Mar 25 12:22 ch_23Aug2018_HeLa_Std_1.raw.mgf
-rw-r--r-- 1 root root   2558440 Mar 25 12:25 qExactive01819.raw_Matrix.txt
-rw-r--r-- 1 root root       733 Mar 25 12:25 qExactive01819.raw_Metrics.txt
-rw-r--r-- 1 root root  29672634 Mar 25 12:25 qExactive01819.raw.mgf
```

**Comment by CarlosHorro** [comment]
The delay with the tests is mainly due to the work necessary to obtain a .raw file as small as possible to be included into the tool for executing tests.
Some days ago I got a .raw file from my lab of about 70mb, which is about 7mb when converted to mgf. Are these sizes (about 80mb in total) valid to be included into the tool @bgruening ?

**Comment by kevinkovalchik** [comment]
Sorry for the delay on the small files. I've had a lot on my plate and have
been putting it off. I should be able to generate some files smaller than
70 Mb that will work with RawTools. Probably in the 10-20 Mb range. I'll
look into it again today.

Kevin

**Comment by bgruening** [comment]
@CarlosHorro :( 70MB are something like 69MB too much. We try to keep test data really small to not waste the space on github, the toolshed and all our deployments.

I guess there is no other way around? If so we will hack this into travis. We will download the big test data from somewhere during testing ... huge hack but hopefully just for 1-2 tools.

**Comment by CarlosHorro** [comment]
> @CarlosHorro :( 70MB are something like 69MB too much. We try to keep test data really small to not waste the space on github, the toolshed and all our deployments.
> 
> I guess there is no other way around? If so we will hack this into travis. We will download the big test data from somewhere during testing ... huge hack but hopefully just for 1-2 tools.

Hi,

@kevinkovalchik has done a great job and got a 1mb raw file which seems to work with RawTools :-)

**Comment by bgruening** [comment]
Awesome!!! Let me know when you need a final review!

**Comment by CarlosHorro** [comment]
Hi,

everything seems to be ready and working, apart from what I asked in
https://github.com/galaxyproteomics/tools-galaxyp/pull/355#issuecomment-476165225
related to the possibility of managing more than 1 file per execution.

**Comment by bgruening** [comment]
can you try `pattern="(?P&lt;designation&gt;.+)\.mgf"`. Does that work?

**Comment by CarlosHorro** [comment]
> can you try `pattern="(?P&lt;designation&gt;.+)\.mgf"`. Does that work?

Thank you very much @bgruening ! It has work magically! :-)

File: tools/rawtools/rawtools.xml [hunk 7]
```diff
@@ -29,7 +29,14 @@
         ## TODO: To remove this patch when dropping support for Galaxy 17_09, adding profile="18.01" to the tool.
         export HOME=\$cwd  &&
 
-        ln -s '$input_raw' input.raw &&
+        mkdir raws_folder &&
+        #for $input_raw in $input_raws:
+            #set $input_name = $input_raw.element_identifier.split('/')[-1].replace(".raw", "") + ".raw"
+            ln -s -f '${input_raw}' './raws_folder/${input_name}' &&
+            #set $encoded_id = $__app__.security.encode_id($input_raw.id)
```
Comment (bgruening) [inline tools/rawtools/rawtools.xml]: uha, why is this needed?

File: tools/rawtools/rawtools.xml [hunk 8]
```diff
@@ -29,7 +29,14 @@
         ## TODO: To remove this patch when dropping support for Galaxy 17_09, adding profile="18.01" to the tool.
         export HOME=\$cwd  &&
 
-        ln -s '$input_raw' input.raw &&
+        mkdir raws_folder &&
+        #for $input_raw in $input_raws:
+            #set $input_name = $input_raw.element_identifier.split('/')[-1].replace(".raw", "") + ".raw"
```
Comment (bgruening) [inline tools/rawtools/rawtools.xml]: You probably want to use something like that as well:

`#set $bam_name = re.sub('[^\w\-\.]', '_', str($bam.element_identifier))` to get rid of spaces and so on ...

File: tools/rawtools/rawtools.xml (hunk 7 above)
Comment (CarlosHorro) [reply tools/rawtools/rawtools.xml]: Ups, there was some old code here for testing, I've removed it.

File: tools/rawtools/rawtools.xml [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+
+RawTools is written in C# and uses the Thermo RawFileReader library. RawTools is the replacement for RawQuant Python package.
+
+https://github.com/kevinkovalchik/RawTools
+
+**License**
+
+Raw Tools is a free open-source project, following an Apache License 2.0.
+
+RawFileReader reading tool. Copyright © 2016 by Thermo Fisher Scientific, Inc. All rights reserved.
```
Comment (bgruening) [inline tools/rawtools/rawtools.xml]: Shit, if this is true we need to include a button as @chambm did for msconvert :(

Comment (CarlosHorro) [reply tools/rawtools/rawtools.xml]: Well it is not a big deal, I have included it :-)

**Comment by CarlosHorro** [comment]
It is quite weird, the error executing test 2 :

`JobOutputsError: 'ascii' codec can't encode character u'\ufeff' in position 78: ordinal not in range(128)`

has been happening in my local for many days, but it did not happen here until now (I just supposed I had something misconfigured in my planemo)... I have not been able to find out how to solve it...

**Comment by bgruening** [comment]
`The Unicode character U+FEFF is the byte order mark, or BOM, and is used to tell the difference between big- and little-endian UTF-16 encoding.` mh ...

File: tools/rawtools/rawtools.xml [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+
+        <!-- Test mgf conversion -->
+        <test>
+            <param name="input_raws" value="really_small.raw" ftype="thermo.raw"/>
+            <param name="license_agreement" value="true" />
+            <param name="pq_conditional|matrix_select" value="no"/>
+            <param name="mgf_conditional|mgf_select" value="yes"/>
+            <param name="mgf_conditional|mgf_refinement" value="yes"/>
+            <output_collection name="output_mgf" type="list">
+                <element name="really_small.raw" file="really_small.raw.mgf" ftype="mgf">
```
Comment (bgruening) [inline tools/rawtools/rawtools.xml]: > file="really_small.raw.mgf" 

can be removed?

File: tools/rawtools/rawtools.xml [hunk 11]
```diff
@@ -94,6 +94,9 @@
     <inputs>
         <param format="thermo.raw" name="input_raws" type="data" label="Raw file" multiple="true"
             help="Select the raw file from history"/>
+        <param name="license_agreement" type="boolean" label="Do you agree to the vendor licenses?" help="This tool uses proprietary vendor libraries; to run it you must agree to the vendor licenses. Read them at http://proteowizard.sourceforge.net/licenses/thermo.html">
+            <validator type="expression" message="You must agree to the vendor licenses to run msconvert.">True == value</validator>
```
Comment (chambm) [inline tools/rawtools/rawtools.xml]: You'll probably want to change the validator message and label to not point at ProteoWizard/msconvert. :) But having this validation on the client side is really nice. I wish more input validation could happen on the client side to avoid failed jobs.

File: tools/rawtools/rawtools.xml [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+
+        2>> $temp_stderr) &&
+
+        cat $temp_stderr 2>&1;
+]]>
+    </command>
+    <inputs>
+        <param format="thermo.raw" name="input_raws" type="data" label="Raw file" multiple="true"
+            help="Select the raw file from history"/>
+        <param name="license_agreement" type="boolean" label="Do you agree to the vendor licenses?" help="This tool uses a proprietary vendor library; to run it you must agree to the vendor license. Read it at http://proteowizard.sourceforge.net/licenses/thermo.html">
```
Comment (bgruening) [inline tools/rawtools/rawtools.xml]: I'm not sure the link is now correct, do you have a better link?

File: tools/rawtools/rawtools.xml (hunk 11 above)
Comment (bgruening) [reply tools/rawtools/rawtools.xml]: There are a lot of these validators and new once can be written :)

File: tools/rawtools/rawtools.xml (hunk 12 above)
Comment (kevinkovalchik) [reply tools/rawtools/rawtools.xml]: That link leads to a license for MSFileReader, but we are using RawFileReader. Here is a better one: https://github.com/kevinkovalchik/RawTools/blob/master/RawFileReaderLicense
I can change it to .md or .txt if needed.

**Comment by bgruening** [comment]
Mh ... one other guess, can it be that on stderr or stdout is some UTF symbol or anything like that only if you run the tool with specific settings ...

**Comment by CarlosHorro** [comment]
> Mh ... one other guess, can it be that on stderr or stdout is some UTF symbol or anything like that only if you run the tool with specific settings ...

I don't know... whatever it is , it seems to be related to the test platform as the mgf generation works well (with the same options) using the tool interface. I don't see either any suspicious character in the stderr or stdout when executing it...

I've also being testing some things I've seen across internet like:

```
#import sys
#reload(sys)
#sys.setdefaultencoding('utf-8')
#set $myencoding = sys.getdefaultencoding()
echo "myencoding: " &&
echo $myencoding &&

```
By default it is "ascii" and I wanted to try changing it to utf-8, but it seems something fails when setting the default encoding because all tests return empty collections:

`Failed to find identifier [really_small] for testing, tool generated collection elements [[]]
`

**Comment by CarlosHorro** [comment]
As I don`t find any way to avoid this problem (which seems to be specific of the testing functionality), I think I am going to remove the second test.

**Comment by bgruening** [comment]
@CarlosHorro the error is actually a valid one just cryptic.

I only get these results:

```
bag@bag:~/projects/code/tools-galaxyp/tools/rawtools$ grep PEPMASS /tmp/tmpc6clwG/job_working_directory/000/2/working/output/really_small.raw.mgf 
PEPMASS=675,248779296875
PEPMASS=407,919311523438
PEPMASS=436,341217041016
PEPMASS=594,222717285156
PEPMASS=511,47119140625
PEPMASS=834,349731445313
PEPMASS=741,7568359375
PEPMASS=899,80419921875
PEPMASS=667,705932617188
PEPMASS=860,806823730469
PEPMASS=905,804077148438
PEPMASS=791,7453928975
PEPMASS=1024,85485839844
PEPMASS=661,706604003906
PEPMASS=1088,44177246094
PEPMASS=915,798645019531
PEPMASS=1018,85278320313
PEPMASS=786,755065917969
PEPMASS=965,25830078125
PEPMASS=853,323913574219
PEPMASS=1137,90335188188
PEPMASS=887,262299635781
PEPMASS=803,747468092812
PEPMASS=1056,93298339844
PEPMASS=623,21337890625
PEPMASS=694,163146972656
PEPMASS=678,697814941406
PEPMASS=630,706604003906
PEPMASS=979,858063795937
PEPMASS=686,680358886719
PEPMASS=805,729125976563
PEPMASS=960,893493652344
PEPMASS=824,2004710225
PEPMASS=1072,44873046875
```
So your asserts are being false. Also please note the `,` instead of a `.`. So your tools seems to take the locale settings into account, which is not good I think.
I tried adding `LC_ALL=C` to the tool, but this did not work.

**Comment by kevinkovalchik** [comment]
Ah. @CarlosHorro are you using v1.4.1? In 1.4.2 we made a change to ignore local culture settings, to prevent this exact issue.

**Comment by bgruening** [comment]
https://github.com/bioconda/bioconda-recipes/pull/14335

**Comment by bgruening** [comment]
@CarlosHorro I think you probably need to adopt the asserts.
@kevinkovalchik thanks!!!

**Comment by CarlosHorro** [comment]
Thank you very much! :-)  I was quite lost with this last problem but it must work now :-)

**Comment by bgruening** [comment]
Awesome!

**Comment by bgruening** [comment]
@CarlosHorro, @kevinkovalchik: https://github.com/usegalaxy-eu/usegalaxy-eu-tools/pull/171

**Comment by CarlosHorro** [comment]
Hi @bgruening ,

Something weird is happening when installing Rawtools from galaxyp. The installation of the tool in Galaxy seems to go well, but running it just returns empty collections or execution never stops.
When looking into the package details, I see the dependency has not been properly managed:

Dependency | Version | Resolver | Exact version | Current Installation Status |  
rawtools | None | None | True |  

ie, it does not detect even we manage a Conda dependency, but the installation status says "True".

And, when executing the tool, I see this in the logs:
```
galaxy.jobs.output_checker INFO 2019-04-05 13:47:52,250 Job 822: Fatal error: Exit code 127 ()
galaxy.jobs.output_checker DEBUG 2019-04-05 13:47:52,250 Tool exit code indicates an error, failing job.
```

Do you know which could be the reason for not detecting the Conda dependency?
Greetings,
Carlos

**Comment by CarlosHorro** [comment]
Well I think I found it:

```
galaxy.tools.deps.conda_util DEBUG 2019-04-05 14:08:04,725 Executing command: /home/parallels/GitHub/galaxy/database/dependencies/_conda/bin/conda create -y --override-channels --channel iuc --channel bioconda --channel conda-forge --channel defaults --channel r --name __rawtools@1.4.2 rawtools=1.4.2
Fetching package metadata .................
Solving package specifications: 

PackageNotFoundError: Packages missing in current channels:
            
  - rawtools 1.4.2* -> mono >=5.0.0 -> libgcc-ng >=7.3.0
  - rawtools 1.4.2* -> mono >=5.0.0 -> libstdcxx-ng >=7.3.0

We have searched for the packages in the following channels:
            
  - https://conda.anaconda.org/iuc/linux-64
  - https://conda.anaconda.org/iuc/noarch
  - https://conda.anaconda.org/bioconda/linux-64
  - https://conda.anaconda.org/bioconda/noarch
  - https://conda.anaconda.org/conda-forge/linux-64
  - https://conda.anaconda.org/conda-forge/noarch
  - https://repo.continuum.io/pkgs/free/linux-64
  - https://repo.continuum.io/pkgs/free/noarch
  - https://repo.continuum.io/pkgs/r/linux-64
  - https://repo.continuum.io/pkgs/r/noarch
  - https://repo.continuum.io/pkgs/pro/linux-64
  - https://repo.continuum.io/pkgs/pro/noarch
  - https://conda.anaconda.org/r/linux-64
  - https://conda.anaconda.org/r/noarch
            

galaxy.tools.deps.resolvers.conda DEBUG 2019-04-05 14:08:15,680 Removing failed conda install of [CondaTarget[rawtools,version=1.4.2]]

```

**Comment by bgruening** [comment]
I think your channel order is incorrect. Which Galaxy version is that?
It's wired since it worked here and was installed correctly?

**Comment by CarlosHorro** [comment]
> I think your channel order is incorrect. Which Galaxy version is that?
> It's wired since it worked here and was installed correctly?

It`s 17.09

**Comment by bgruening** [comment]
Uha ... can you try to update your conda version? What is your conda version.

**Comment by CarlosHorro** [comment]
> Uha ... can you try to update your conda version? What is your conda version.

Yep I am working on that... actual one is 4.5.12...

**Comment by CarlosHorro** [comment]
After updating conda and stablishing conda channels like here:

https://github.com/bioconda/bioconda-recipes/blob/master/config.yml

it did not work either, and I had to add "anaconda" channel like this:
conda_ensure_channels = conda-forge,bioconda,defaults,r,anaconda

in order it properly gets both packages from the anaconda channel :-)

Thank you very much,
Carlos

**Comment by bgruening** [comment]
wow, that is wired, anaconda == defaults ... but glad that it is working ... But hey ... update!!!! :)

Carlos ... it would be great if you could join us at https://galaxyproject.org/events/gcc2019/!

**Comment by CarlosHorro** [comment]
> wow, that is wired, anaconda == defaults ... but glad that it is working ... But hey ... update!!!! :)
> 
> Carlos ... it would be great if you could join us at https://galaxyproject.org/events/gcc2019/!

I can assure you I would really like to update our version hehe, but it's quite complicated until we finish the current development we are working on...

I would also love to go to the gcc but I'm afraid it's currently impossible... there will be a new member in my family very soon ;-) and I'm going to be extra-busy for the next months :-)
